packages/@aws-cdk/cdk/lib/resource.ts

```diff
@@ -35,12 +37,48 @@ export interface ResourceProps {
 export abstract class Resource extends Construct implements IResource {
   public readonly stack: Stack;
 
-  private readonly _physicalName: PhysicalName;
+  /**
+   * Returns a string-encoded token that resolves to the physical name that should be
+   * passed to the CloudFormation resource.
+   *
+   * This token will resolve to one of:
+   * - a concrete value
+   * - `undefined`, when a name should be generated by cloudformation
+   * - a name generated automatically during synthesis, in cross-account scenario.
+   *
+   * @experimental
+   */
+  protected readonly physicalName: string;
+
+  private _physicalName: string | undefined;
+  private readonly _allowCrossEnvironment: boolean;
 
   constructor(scope: Construct, id: string, props: ResourceProps = {}) {
     super(scope, id);
     this.stack = Stack.of(this);
-    this._physicalName = props.physicalName || PhysicalName.auto();
+
+    let physicalName = props.physicalName;
+
+    if (props.physicalName === PhysicalName.GENERATE_IF_NEEDED) {
```

Comment:
=== check will fail if we have multiple copies of the lib. People still manage to do this apparently, see the issue for the "store traits on objects" pr that Romain did a couple of days ago

Reply:
Good call. What would you recommend instead?

Reply:
I'll leave it like this for now and we'll come up with something at a later stage.